File: 06-inverted-binary-tree/06-invert-binary-tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __str__(self):
        lines = [self.__class__.__name__ + ':']
        for key, val in vars(self).items():
            lines += '{}: {}'.format(key, val).split('\n')
        return '\n    '.join(lines)

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        aux = None
        if (root):
            if (root.left):
                root.left = self.invertTree(root.left)
            if (root.right):
                root.right = self.invertTree(root.right)

            aux = root.left
            root.left = root.right
            root.right = aux
        return root


print(Solution().invertTree(TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(7, TreeNode(6), TreeNode(9)))))

File: 06-inverted-binary-tree/06-invert-binary-tree.js
// https://leetcode.com/problems/invert-binary-tree/
class TreeNode {
    val;
    left;
    right;
    constructor(val, left, right) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
const invertTree = function (root) {
    let aux;
    if (root) {
        if (root.left) {
            root.left = invertTree(root.left);
        }

        if (root.right) {
            root.right = invertTree(root.right);
        }
        aux = root.left;
        root.left = root.right;
        root.right = aux;
    }
    return root;
};

// if root has a left , call recursively to root.left
// if root has a right, call recursively to root.right
// execute inversion and return node
console.log(
    invertTree(
        new TreeNode(
            4,
            new TreeNode(7, new TreeNode(6), new TreeNode(9)),
            new TreeNode(2, new TreeNode(1), new TreeNode(3))
        )
    )
);

//           4
//      2          7
//    1    3     6  9
//   1 2 4 7   9 2 0 1
